File: server/src/graphs/graphs.service.ts
import { CACHE_MANAGER, Inject, Injectable, Logger, OnApplicationBootstrap } from '@nestjs/common';
import { readFileSync } from 'fs';
import { join } from 'path';
import * as xml2js from 'xml2js';
import { Cache } from 'cache-manager';

@Injectable()
export class GraphsService implements OnApplicationBootstrap {
  constructor(@Inject(CACHE_MANAGER) private cacheManager: Cache) {}

  /**
   * Builds a graph data object from the cached XML configuration and returns it as a JSON string.
   */
  async getGraphData(serverType: string): Promise<string> {
    const cache = await this.getFromRedisXMLCache(serverType);
    // Check if cache exists and create if not.
    if (!cache) await this.createRedisXMLCache(serverType);
    // Get from redis cache
    return this.getFromRedisXMLCache(serverType).then(async (result) => {
      const gdata = process.env['ENV'] == 'PROD' ? await this.getFromRedisGraphDataCache(serverType) : null;

      if (!gdata) await this.createRedisGraphDataCache(serverType);

      return gdata ? gdata : this.getFromRedisGraphDataCache(serverType);
    });
  }

  /**
   * Forces the redis cache to rebuild.
   */
  async getGraphDataForceRebuild(serverType: string, from = 'Client') {
    Logger.warn(`${from} forced a reload`);
    await this.createRedisXMLCache(serverType);
    // Get from redis cache
    return this.getFromRedisXMLCache(serverType).then(async (result) => {
      await this.createRedisGraphDataCache(serverType);

      return this.getFromRedisGraphDataCache(serverType);
    });
  }

  /**
   * Creates a new graph data cache.
   */
  async createRedisGraphDataCache(serverType: string) {
    await this.cacheManager.set(
      `serverConfiguration:gdata:${serverType}`,
      await this.buildGraphDataWithXMLCache(serverType),
      { ttl: 43200 }, // 12 hours
    );

    Logger.log('Created new gdata cache.');
  }

  /**
   * Retrieves graph data from the redis cache.
   */
  async getFromRedisGraphDataCache(serverType: string) {
    return JSON.parse(await this.cacheManager.get(`serverConfiguration:gdata:${serverType}`));
  }

  /**
   * Creates a copy of the parsed XML file in memory
   */
  async createRedisXMLCache(serverType: string) {
    await this.cacheManager.set(
      `serverConfiguration:xml:${serverType}`,
      JSON.stringify(await this.getData(process.env[serverType])),
      { ttl: 43200 }, // 12 hours
    );

    Logger.log('Created new XML cache.');
  }

  /**
   * Retrieves a copy of the parsed XML file from memory.
   */
  async getFromRedisXMLCache(serverType: string) {
    return JSON.parse(await this.cacheManager.get(`serverConfiguration:xml:${serverType}`));
  }

  /**
   * Weird stuff was going on, so I had to create this function.
   */
  async buildGraphDataWithXMLCache(serverType: string): Promise<string> {
    const cache = await this.getFromRedisXMLCache(serverType);
    // Check if cache exists and create if not.
    if (!cache) {
      await this.createRedisXMLCache(serverType);

      return this.getFromRedisXMLCache(serverType).then((result) => {
        return this.buildGraphData(result);
      });
    } else {
      return this.buildGraphData(cache);
    }
  }

  /**
   * Builds a graph data object (https://github.com/vasturiano/force-graph#input-json-syntax)
   * @param xml
   */
  buildGraphData(xml: any): string {
    // Create empty graph data object
    const gData = {
      nodes: [],
      links: [],
    };

    const OTHER = 'OTHER';

    gData.nodes.push({
      id: OTHER,
      name: OTHER,
      group: OTHER,
      description: 'Unhandled connectors',
      val: 1,
      tags: [],
    });

    // XML structure is a bit weird. <channels> is an array, so it's the 'root' so to speak.
    xml.serverConfiguration.channels.forEach((root) => {
      // Go through each channel
      root.channel.forEach((channel) => {
        // Extract ID and name
        const channelId = channel.id[0];
        const channelName = channel.name[0];
        const channelDescription = channel.description[0];

        // Create a new node from the channel's ID and name
        gData.nodes.push({
          id: channelId,
          name: 'Channel: ' + channelName,
          val: 1,
          description: channelDescription,
          group: channel.exportData[0].metadata[0].enabled[0] == 'true' ? 'Channel' : 'disabled',
          enabled: channel.exportData[0].metadata[0].enabled[0] == 'true' ? 1 : 0,
          tags: [],
        });

        // sourceConnector is an array holding the data source of the channel
        channel.sourceConnector.forEach((sourceConnector) => {
          const sourceConnectorProperties = sourceConnector.properties[0];
          const transportName = sourceConnector.transportName[0];

          const sourceTransformer = sourceConnector.transformer[0];
          const sourceFilter = sourceConnector.filter[0];

          switch (transportName) {
            case 'TCP Listener':
              const tcpListenerConnectorProperties = sourceConnectorProperties.listenerConnectorProperties[0];
              const tcpInnerSourceConnectorProperties = sourceConnectorProperties.sourceConnectorProperties[0];

              const tcpListenerHost = tcpListenerConnectorProperties.host[0];
              const tcpListenerPort = tcpListenerConnectorProperties.port[0];

              const tcpListenerID = `${tcpListenerHost}:${tcpListenerPort}`;

              if (!gData.nodes.find((node) => node.id == tcpListenerID)) {
                gData.nodes.push({
                  id: tcpListenerID,
                  name: `${transportName}: ${tcpListenerID}`,
                  val: 1,
                  group: transportName,
                  tags: [],
                });
              }

              gData.links.push({
                source: tcpListenerID,
                target: channelId,
                group: transportName,
                enabled: sourceConnector.enabled[0] == 'true' ? 1 : 0,
              });

              break;
            case 'HTTP Listener':
              const httpListenerConnectorProperties = sourceConnectorProperties.listenerConnectorProperties[0];
              const httpInnerSourceConnectorProperties = sourceConnectorProperties.sourceConnectorProperties[0];

              const httpListenerHost = httpListenerConnectorProperties.host[0];
              const httpListenerPort = httpListenerConnectorProperties.port[0];

              const httpListenerID = `${httpListenerHost}:${httpListenerPort}`;

              if (!gData.nodes.find((node) => node.id == httpListenerID)) {
                gData.nodes.push({
                  id: httpListenerID,
                  name: `${transportName}: ${httpListenerID}`,
                  val: 1,
                  group: transportName,
                  tags: [],
                });
              }

              gData.links.push({
                source: httpListenerID,
                target: channelId,
                group: transportName,
                enabled: sourceConnector.enabled[0] == 'true' ? 1 : 0,
              });

              break;
            case 'Database Reader':
              const dbReaderID = sourceConnectorProperties.url[0];
              const dbHost = dbReaderID.split(/(@|\/\/)/)[2];

              if (!gData.nodes.find((node) => node.id == dbHost)) {
                gData.nodes.push({
                  id: dbHost,
                  name: `Database Host: ${dbHost}`,
                  group: 'Host',
                  description: `DB Host\nTreiber: ${sourceConnectorProperties.driver[0]}`,
                  val: 1,
                  tags: [],
                });
              }

              if (!gData.nodes.find((node) => node.id == dbReaderID)) {
                gData.nodes.push({
                  id: dbReaderID,
                  name: `${transportName}: ${dbReaderID}`,
                  val: 1,
                  group: transportName,
                  tags: [],
                });
              }

              gData.links.push({
                source: dbHost,
                target: dbReaderID,
                group: transportName,
                enabled: sourceConnector.enabled[0] == 'true' ? 1 : 0,
              });

              gData.links.push({
                source: dbReaderID,
                target: channelId,
                group: transportName,
                enabled: sourceConnector.enabled[0] == 'true' ? 1 : 0,
              });
              break;
            case 'File Reader':
              const fileReaderID = sourceConnectorProperties.host[0];

              if (!gData.nodes.find((node) => node.id == fileReaderID)) {
                gData.nodes.push({
                  id: fileReaderID,
                  name: `${transportName}: ${fileReaderID}`,
                  val: 1,
                  group: transportName,
                  tags: [],
                });
              }

              gData.links.push({
                source: fileReaderID,
                target: channelId,
                group: transportName,
                enabled: sourceConnector.enabled[0] == 'true' ? 1 : 0,
              });
              break;
            case 'DICOM Listener':
              const dicomListenerConnectorProperties = sourceConnectorProperties.listenerConnectorProperties[0];
              const dicomListenerID = `${sourceConnectorProperties.applicationEntity[0]}:${dicomListenerConnectorProperties.port[0]}`;

              if (!gData.nodes.find((node) => node.id == dicomListenerID)) {
                gData.nodes.push({
                  id: dicomListenerID,
                  name: `${transportName}: ${dicomListenerID}`,
                  val: 1,
                  group: transportName,
                  tags: [],
                });
              }

              gData.links.push({
                source: dicomListenerID,
                target: channelId,
                group: transportName,
                enabled: sourceConnector.enabled[0] == 'true' ? 1 : 0,
              });
              break;
            case 'JavaScript Reader':
              break;
          }

          // Catch router.routeMessage(<channelName>, payload) links
          sourceTransformer.elements.forEach((element) => {
            const jsStep = element['com.mirth.connect.plugins.javascriptstep.JavaScriptStep'];

            if (jsStep) {
              addRouterNodesAndLinks(sourceConnector, jsStep);
            }
          });

          sourceFilter.elements.forEach((element) => {
            const jsStep = element['com.mirth.connect.plugins.javascriptstep.JavaScriptStep'];

            if (jsStep) {
              addRouterNodesAndLinks(sourceConnector, jsStep);
            }
          });
        });

        // destinationConnector is an array holding every connector in the channel
        channel.destinationConnectors.forEach((destinationConnector) => {
          // Go through each connector
          destinationConnector.connector.forEach((connector) => {
            const connectorProperties = connector.properties[0];
            const transportName = connector.transportName[0];

            const connectorTransformer = connector.transformer[0];
            const connectorFilter = connector.filter[0];
            const connectorResponseTransformer = connector.responseTransformer[0];

            // Check if connector is enabled
            // Check for different transports
            switch (transportName) {
              // Is a channel writer?
              case 'Channel Writer':
                // Extract target channel id
                const targetChannelId = connectorProperties.channelId[0];

                let isTargetEnabled = false;

                // Check if the target channel is enabled
                root.channel.forEach((targetChannel) => {
                  if (targetChannel.id[0] == targetChannelId) {
                    isTargetEnabled = targetChannel.exportData[0].metadata[0].enabled[0] == 'true';
                  }
                });

                if (isTargetEnabled) {
                  // Create a new link
                  if (targetChannelId != 'none') {
                    gData.links.push({
                      source: channelId,
                      target: targetChannelId,
                      enabled: connector.enabled[0] == 'true' ? 1 : 0,
                      group: 'Channel Writer',
                    });
                  } else {
                    gData.links.push({
                      source: channelId,
                      target: 'other',
                      enabled: connector.enabled[0] == 'true' ? 1 : 0,
                      group: 'Channel Writer',
                    });
                  }
                }

                break;
              case 'SMTP Sender':
                const regex =
                  // eslint-disable-next-line max-len
                  /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

                const to = connectorProperties.to[0].toLowerCase();

                if (to) {
                  const emails = to.split(',');

                  emails.forEach((email) => {
                    email = email.trim();
                    email = email.toLowerCase();

                    if (email.search(regex) != -1) {
                      if (!gData.nodes.find((node) => node.id == email)) {
                        gData.nodes.push({
                          id: email,
                          name: 'SMTP: ' + email,
                          val: 1,
                          group: transportName,
                          tags: [],
                        });
                      }

                      gData.links.push({
                        source: channelId,
                        target: email,
                        group: 'SMTP Sender',
                        enabled: connector.enabled[0] == 'true' ? 1 : 0,
                      });
                    } else {
                      if (!gData.nodes.find((node) => node.id == to)) {
                        gData.nodes.push({
                          id: to,
                          name: 'SMTP: ' + to,
                          val: 1,
                          group: 'Unreadable ' + transportName,
                          tags: [],
                        });
                      }

                      gData.links.push({
                        source: channelId,
                        target: to,
                        enabled: connector.enabled[0] == 'true' ? 1 : 0,
                        group: 'Unreadable ' + transportName,
                      });
                    }
                  });
                }
                break;
              case 'TCP Sender':
                const remoteAddress = connectorProperties.remoteAddress[0];

                if (remoteAddress) {
                  const remoteAddressAndPort = remoteAddress + ':' + connectorProperties.remotePort[0];

                  if (!gData.nodes.find((node) => node.id == remoteAddressAndPort)) {
                    gData.nodes.push({
                      id: remoteAddressAndPort,
                      name: 'TCP Remote: ' + remoteAddressAndPort,
                      group: transportName,
                      val: 1,
                      tags: [],
                    });
                  }

                  gData.links.push({
                    source: channelId,
                    target: remoteAddressAndPort,
                    group: transportName,
                    enabled: connector.enabled[0] == 'true' ? 1 : 0,
                  });
                }
                break;
              case 'File Writer':
                const host = connectorProperties.host[0].toLowerCase();

                if (host) {
                  if (!gData.nodes.find((node) => node.id == host)) {
                    gData.nodes.push({
                      id: host,
                      name: 'File Host: ' + host,
                      group: transportName,
                      val: 1,
                      tags: [],
                    });
                  }

                  gData.links.push({
                    source: channelId,
                    target: host,
                    group: transportName,
                    enabled: connector.enabled[0] == 'true' ? 1 : 0,
                  });
                }
                break;
              default:
                // If the connector is unknown, add a link to OTHER
                // TODO: Currently breaks the graph
                /*gData.links.push({
                  source: channelId,
                  target: OTHER,
                  group: OTHER,
                  enabled: connector.enabled[0] == 'true' ? 1 : 0,
                });*/
                break;
            }

            // Catch router.routeMessage(<channelName>, payload) links
            connectorTransformer.elements.forEach((element) => {
              const jsStep = element['com.mirth.connect.plugins.javascriptstep.JavaScriptStep'];

              if (jsStep) {
                addRouterNodesAndLinks(connector, jsStep);
              }
            });

            connectorFilter.elements.forEach((element) => {
              const jsStep = element['com.mirth.connect.plugins.javascriptstep.JavaScriptStep'];

              if (jsStep) {
                addRouterNodesAndLinks(connector, jsStep);
              }
            });

            connectorResponseTransformer.elements.forEach((element) => {
              const jsStep = element['com.mirth.connect.plugins.javascriptstep.JavaScriptStep'];

              if (jsStep) {
                addRouterNodesAndLinks(connector, jsStep);
              }
            });
          });
        });

        /**
         * Tries to parse JavaScript steps in filters and connectors, then looks for manual routing
         * @param connector Connector XML node
         * @param jsStep JavaScript step XML node
         */
        function addRouterNodesAndLinks(connector, jsStep) {
          /**
           * Uses RegEx to find the target of a given manual routing attempt, then returns an object containing either
           */
          function getChannelNameOrID(): { channelName: string | undefined; channelId: string | undefined } {
            //console.log(jsStep[0].script[0]);

            const regexName = /^\s*router\.routeMessage\(['"](\w+)['"],/;
            const regexID = /^\s*router\.routeMessageByChannelId\(['"]([\w\-]+)['"],/;

            // Split JS step into single lines
            const scriptSplit = jsStep[0].script[0].split('\n');

            let channelName;
            let channelId;

            // Go through each line
            scriptSplit.forEach((str: string) => {
              let nameRegexResult;
              let idRegExResult;

              // Try to find router.routeMessage()
              if ((nameRegexResult = regexName.exec(str))) {
                // This is necessary to avoid infinite loops with zero-width matches
                if (nameRegexResult.index === regexName.lastIndex) {
                  regexName.lastIndex++;
                }

                // Get from index 1, because the RegEx contains a capture group
                channelName = nameRegexResult[1];
              }

              // Try to find router.routeMessageByChannelId()
              if ((idRegExResult = regexID.exec(str))) {
                // This is necessary to avoid infinite loops with zero-width matches
                if (idRegExResult.index === regexID.lastIndex) {
                  regexID.lastIndex++;
                }

                // Get from index 1, because the RegEx contains a capture group
                channelId = idRegExResult[1];
              }
            });

            return {
              channelName: channelName,
              channelId: channelId,
            };
          }

          const result = getChannelNameOrID();

          // console.log(result);

          // If result contains the target channel's name
          if (result.channelName) {
            // Find the channel in the XML structure
            root.channel.forEach((targetChannel) => {
              if (targetChannel.name[0] == result.channelName) {
                const nodeResult = gData.nodes.find((node) => (node.id = targetChannel.id[0]));

                // If gData doesn't contain the node yet, add it
                if (!nodeResult) {
                  gData.nodes.push({
                    id: targetChannel.id[0],
                    name: 'Router Target: ' + targetChannel.name[0],
                    val: 1,
                    description: targetChannel.description[0],
                    group: 'Router Target',
                    enabled: targetChannel.exportData[0].metadata[0].enabled[0] == 'true' ? 1 : 0,
                    tags: [],
                  });
                }

                // Add custom router link
                gData.links.push({
                  source: channelId,
                  target: targetChannel.id[0],
                  enabled: connector.enabled[0] == 'true' ? 1 : 0,
                  group: 'Router',
                });
              }
            });
          } else if (result.channelId) {
            // If result contains ID, but not the name
            const nodeResult = gData.nodes.find((node) => (node.id = result.channelId));

            // If gData doesn't contain the node yet, add it
            if (!nodeResult) {
              root.channel.forEach((targetChannel) => {
                if (targetChannel.id[0] == result.channelId) {
                  gData.nodes.push({
                    id: targetChannel.id[0],
                    name: 'Router Target: ' + targetChannel.name[0],
                    val: 1,
                    group: 'Router Target',
                    tags: [],
                  });
                }
              });
            }

            // Add custom router link
            gData.links.push({
              source: channelId,
              target: result.channelId,
              enabled: connector.enabled[0] == 'true' ? 1 : 0,
              group: 'Router',
            });
          }
        }
      });
    });

    xml.serverConfiguration.channelTags.forEach((root) => {
      root.channelTag.forEach((channelTag) => {
        const tagName = channelTag.name[0];

        channelTag.channelIds.forEach((channelIdRoot) => {
          if (channelIdRoot.string) {
            channelIdRoot.string.forEach((channelId) => {
              xml.serverConfiguration.channels.forEach((channelRoot) => {
                // Go through each channel
                channelRoot.channel.forEach((channel) => {
                  if (channel.id[0] == channelId) {
                    const result = gData.nodes.find((node) => node.id == channelId);
                    result.tags.push(tagName);
                  }
                });
              });
            });
          }
        });
      });
    });

    return JSON.stringify(gData);
  }

  /**
   * Parses the backup file as XML and converts it into a JSO.
   */
  getData(fileName: string): Promise<any> {
    const file = readFileSync(join('public', fileName));

    return xml2js
      .parseStringPromise(file)
      .then((result) => {
        return result;
      })
      .catch((error) => {
        Logger.error(error);
      });
  }

  async onApplicationBootstrap(): Promise<any> {
    if (process.env.ENV == 'PROD') {
      await this.getGraphDataForceRebuild('DATA_PRODUCTION', 'Startup');
      await this.getGraphDataForceRebuild('DATA_DICOM', 'Startup');
      await this.getGraphDataForceRebuild('DATA_TEST', 'Startup');
    }
  }
}

File: mirthgraph/Services/GraphsService.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Serialization;
using StackExchange.Redis;
using System.Xml.Linq;

// Service for retrieving and processing graph data from Mirth Connect configurations.
public class GraphsService
{
    private readonly IConnectionMultiplexer _redis; // Redis connection multiplexer
    private readonly ILogger<GraphsService> _logger; // Logger for logging messages
    private readonly IDatabase _db; // Redis database instance
    private readonly MirthConfigService _mirthConfigService; // Service for retrieving Mirth Connect configurations

    // Constructor for the GraphsService class.
    public GraphsService(IConnectionMultiplexer redis, ILogger<GraphsService> logger, MirthConfigService mirthConfigService)
    {
        _redis = redis;
        _logger = logger;
        _db = _redis.GetDatabase();
        _mirthConfigService = mirthConfigService;
    }

    // Retrieves graph data for a given connection name.  Checks the cache first.
    public async Task<string> GetGraphDataAsync(string connectionName)
    {
        // Check if graph data is already cached.
        var cachedData = await _db.StringGetAsync($"graphData:{connectionName}");
        if (!cachedData.IsNullOrEmpty)
        {
            // Return cached data if available.
            return cachedData;
        }

        // Build graph data if not cached.
        var graphData = await BuildGraphDataAsync(connectionName);
        // Cache the built graph data.
        await _db.StringSetAsync($"graphData:{connectionName}", graphData);
        return graphData;
    }

    // Builds graph data from a Mirth Connect configuration.
    public async Task<string> BuildGraphDataAsync(string connectionName)
    {
        // Retrieve Mirth Connect configuration for the given connection name.
        var xmlConfig = await _mirthConfigService.GetMirthConfigAsync(connectionName);
        // Parse the XML configuration to extract graph data.
        var graphData = ParseMirthConfig(xmlConfig);
        // Configure JSON serializer settings for camel case naming.
        var serializerSettings = new JsonSerializerSettings();
        serializerSettings.ContractResolver = new DefaultContractResolver
        {
            NamingStrategy = new CamelCaseNamingStrategy
            {
                ProcessDictionaryKeys = false,
                OverrideSpecifiedNames = true
            }
        };
        // Serialize the graph data to JSON.
        var json = JsonConvert.SerializeObject(graphData, Formatting.Indented, serializerSettings);

        // _logger.LogDebug($"Graph data for {connectionName}:\n{json}"); // Uncomment for debugging

        return json;
    }

    // Parses Mirth Connect XML configuration and extracts graph data.
    private GraphData ParseMirthConfig(string xmlString)
    {
        // Create a new GraphData object to store the parsed data.
        var graphData = new GraphData
        {
            Nodes = new List<Node>(),
            Links = new List<Link>()
        };

        // Add a default node for unhandled connectors.
        const string OTHER = "OTHER";

        graphData.Nodes.Add(new Node
        {
            Id = OTHER,
            Name = OTHER,
            Group = OTHER,
            Description = "Unhandled connectors",
            Val = 1,
            Tags = new List<string>()
        });

        // Parse the XML string into an XDocument.
        var xml = XDocument.Parse(xmlString);
        // Select all channel elements from the XML document.
        var channels = xml.Descendants("channel");

        // Iterate over each channel element.
        foreach (var channel in channels)
        {
            // Extract channel information.
            var channelId = channel.Element("id")?.Value;
            var channelName = channel.Element("name")?.Value;
            var channelDescription = channel.Element("description")?.Value;

            // Check if channel ID is available.
            if (channelId != null)
            {
                // Add a node for the channel.
                graphData.Nodes.Add(new Node
                {
                    Id = channelId,
                    Name = "Channel: " + channelName,
                    Val = 1,
                    Description = channelDescription,
                    Group = channel.Element("exportData")?.Element("metadata")?.Element("enabled")?.Value == "true" ? "Channel" : "disabled",
                    Enabled = channel.Element("exportData")?.Element("metadata")?.Element("enabled")?.Value == "true" ? 1 : 0,
                    Tags = new List<string>()
                });

                // Process source connectors for the channel.
                var sourceConnectors = channel.Elements("sourceConnector");
                foreach (var sourceConnector in sourceConnectors)
                {
                    var sourceConnectorProperties = sourceConnector.Element("properties");
                    var transportName = sourceConnector.Element("transportName")?.Value;

                    // Handle different transport types.
                    switch (transportName)
                    {
                        case "TCP Listener":
                            TcpListenerNodeAndLink(graphData, channelId, sourceConnectorProperties, sourceConnector);
                            break;

                        case "HTTP Listener":
                            HttpListenerNodeAndLink(graphData, channelId, sourceConnectorProperties, sourceConnector);
                            break;

                        case "Database Reader":
                            DatabaseReaderNodeAndLink(graphData, channelId, sourceConnectorProperties, sourceConnector);
                            break;

                        case "File Reader":
                            FileReaderNodeAndLink(graphData, channelId, sourceConnectorProperties, sourceConnector);
                            break;

                        case "DICOM Listener":
                            DicomListenerNodeAndLink(graphData, channelId, sourceConnectorProperties, sourceConnector);
                            break;

                        default:
                            // Handle other transport types if necessary
                            break;
                    }
                }
            }
        }

        return graphData;
    }

    // Adds a TCP Listener node and link to the graph data.
    private void TcpListenerNodeAndLink(GraphData graphData, string channelId, XElement sourceConnectorProperties, XElement sourceConnector)
    {
        // Extract TCP Listener properties.
        var tcpListenerConnectorProperties = sourceConnectorProperties?.Element("listenerConnectorProperties");
        var tcpListenerHost = tcpListenerConnectorProperties?.Element("host")?.Value;
        var tcpListenerPort = tcpListenerConnectorProperties?.Element("port")?.Value;
        var tcpListenerID = $"{tcpListenerHost}:{tcpListenerPort}";

        // Add TCP Listener node if it doesn't exist.
        if (!graphData.Nodes.Any(node => node.Id == tcpListenerID))
        {
            graphData.Nodes.Add(new Node
            {
                Id = tcpListenerID,
                Name = $"TCP Listener: {tcpListenerID}",
                Val = 1,
                Group = "TCP Listener",
                Tags = new List<string>()
            });
        }

        // Add a link from the TCP Listener to the channel.
        graphData.Links.Add(new Link
        {
            Source = tcpListenerID,
            Target = channelId,
            Group = "TCP Listener",
            Enabled = sourceConnector.Element("enabled")?.Value == "true" ? 1 : 0
        });
    }

    // Adds an HTTP Listener node and link to the graph data.
    private void HttpListenerNodeAndLink(GraphData graphData, string channelId, XElement sourceConnectorProperties, XElement sourceConnector)
    {
        // Extract HTTP Listener properties.
        var httpListenerConnectorProperties = sourceConnectorProperties?.Element("listenerConnectorProperties");
        var httpListenerHost = httpListenerConnectorProperties?.Element("host")?.Value;
        var httpListenerPort = httpListenerConnectorProperties?.Element("port")?.Value;
        var httpListenerID = $"{httpListenerHost}:{httpListenerPort}";

        // Add HTTP Listener node if it doesn't exist.
        if (!graphData.Nodes.Any(node => node.Id == httpListenerID))
        {
            graphData.Nodes.Add(new Node
            {
                Id = httpListenerID,
                Name = $"HTTP Listener: {httpListenerID}",
                Val = 1,
                Group = "HTTP Listener",
                Tags = new List<string>()
            });
        }

        // Add a link from the HTTP Listener to the channel.
        graphData.Links.Add(new Link
        {
            Source = httpListenerID,
            Target = channelId,
            Group = "HTTP Listener",
            Enabled = sourceConnector.Element("enabled")?.Value == "true" ? 1 : 0
        });
    }

    // Adds a Database Reader node and link to the graph data.
    private void DatabaseReaderNodeAndLink(GraphData graphData, string channelId, XElement sourceConnectorProperties, XElement sourceConnector)
    {
        // Extract Database Reader properties.
        var dbReaderID = sourceConnectorProperties?.Element("url")?.Value;
        var dbHost = dbReaderID?.Split(new[] { '@', '/' }, StringSplitOptions.RemoveEmptyEntries).ElementAtOrDefault(1);

        // Add Database Host node if it doesn't exist.
        if (!graphData.Nodes.Any(node => node.Id == dbHost))
        {
            graphData.Nodes.Add(new Node
            {
                Id = dbHost,
                Name = $"Database Host: {dbHost}",
                Group = "Host",
                Description = $"DB Host\nDriver: {sourceConnectorProperties?.Element("driver")?.Value}",
                Val = 1,
                Tags = new List<string>()
            });
        }

        // Add Database Reader node if it doesn't exist.
        if (!graphData.Nodes.Any(node => node.Id == dbReaderID))
        {
            graphData.Nodes.Add(new Node
            {
                Id = dbReaderID,
                Name = $"Database Reader: {dbReaderID}",
                Val = 1,
                Group = "Database Reader",
                Tags = new List<string>()
            });
        }

        // Add links from the Database Host to the Database Reader and from the Database Reader to the channel.
        graphData.Links.Add(new Link
        {
            Source = dbHost,
            Target = dbReaderID,
            Group = "Database Reader",
            Enabled = sourceConnector.Element("enabled")?.Value == "true" ? 1 : 0
        });

        graphData.Links.Add(new Link
        {
            Source = dbReaderID,
            Target = channelId,
            Group = "Database Reader",
            Enabled = sourceConnector.Element("enabled")?.Value == "true" ? 1 : 0
        });
    }

    // Adds a File Reader node and link to the graph data.
    private void FileReaderNodeAndLink(GraphData graphData, string channelId, XElement sourceConnectorProperties, XElement sourceConnector)
    {
        // Extract File Reader properties.
        var fileReaderPath = sourceConnectorProperties?.Element("host")?.Value;

        // Add File Reader node if it doesn't exist.
        if (!graphData.Nodes.Any(node => node.Id == fileReaderPath))
        {
            graphData.Nodes.Add(new Node
            {
                Id = fileReaderPath,
                Name = $"File Reader: {fileReaderPath}",
                Val = 1,
                Group = "File Reader",
                Tags = new List<string>()
            });
        }

        // Add a link from the File Reader to the channel.
        graphData.Links.Add(new Link
        {
            Source = fileReaderPath,
            Target = channelId,
            Group = "File Reader",
            Enabled = sourceConnector.Element("enabled")?.Value == "true" ? 1 : 0
        });
    }

    // Adds a DICOM Listener node and link to the graph data.
    private void DicomListenerNodeAndLink(GraphData graphData, string channelId, XElement sourceConnectorProperties, XElement sourceConnector)
    {
        // Extract DICOM Listener properties.
        var dicomListenerConnectorProperties = sourceConnectorProperties?.Element("listenerConnectorProperties");
        var dicomListenerID = $"{sourceConnectorProperties?.Element("applicationEntity")?.Value}:{dicomListenerConnectorProperties?.Element("port")?.Value}";

        // Add DICOM Listener node if it doesn't exist.
        if (!graphData.Nodes.Any(node => node.Id == dicomListenerID))
        {
            graphData.Nodes.Add(new Node
            {
                Id = dicomListenerID,
                Name = $"DICOM Listener: {dicomListenerID}",
                Val = 1,
                Group = "DICOM Listener",
                Tags = new List<string>()
            });
        }

        // Add a link from the DICOM Listener to the channel.
        graphData.Links.Add(new Link
        {
            Source = dicomListenerID,
            Target = channelId,
            Group = "DICOM Listener",
            Enabled = sourceConnector.Element("enabled")?.Value == "true" ? 1 : 0
        });
    }
}

// Represents the graph data structure.
public class GraphData
{
    public List<Node> Nodes { get; set; }
    public List<Link> Links { get; set; }
}

// Represents a node in the graph.
public class Node
{
    public string Id { get; set; }
    public string Name { get; set; }
    public string Group { get; set; }
    public string Description { get; set; }
    public int Val { get; set; }
    public int Enabled { get; set; }
    public List<string> Tags { get; set; }
}

// Represents a link between nodes in the graph.
public class Link
{
    public string Source { get; set; }
    public string Target { get; set; }
    public string Group { get; set; }
    public int Enabled { get; set; }
}
